controller/api/destination/endpoint_listener.go

```diff
@@ -131,29 +134,37 @@ func (l *endpointListener) SetServiceID(id *serviceID) {
 	}
 }
 
+// Update is called with lists of newly added and/or removed pods in a service
+// and address updates on the listener's gRPC response stream.
 func (l *endpointListener) Update(add, remove []*updateAddress) {
-	l.log.Debugf("Update(%+v, %+v)", add, remove)
+	l.log.Debugf("Update: add=%d; remove=%d", len(add), len(remove))
 
 	if len(add) > 0 {
-		update := &pb.Update{
-			Update: &pb.Update_Add{
-				Add: l.toWeightedAddrSet(add),
-			},
+		// If pods were added, send the list of metadata-rich WeightedAddr endpoints.
+		set := &pb.WeightedAddrSet{MetricLabels: l.labels}
+		for _, a := range add {
+			w := l.toWeightedAddr(a)
+			l.log.Debugf("Update: add: pod=%s; addr=%s; %+v", a.pod.Name, a.pod.Status.PodIP, w)
+			set.Addrs = append(set.Addrs, w)
 		}
-		err := l.stream.Send(update)
-		if err != nil {
-			l.log.Error(err)
+
+		u := &pb.Update{Update: &pb.Update_Add{Add: set}}
+		if err := l.stream.Send(u); err != nil {
+			l.log.Errorf("Failed to send address update: %s", err)
 		}
 	}
+
 	if len(remove) > 0 {
-		update := &pb.Update{
-			Update: &pb.Update_Remove{
-				Remove: l.toAddrSet(remove),
-			},
+		// If pods were removed, send the list of IP addresses.
+		set := &pb.AddrSet{}
+		for _, a := range remove {
+			l.log.Debugf("Update: remove: pod=%s; addr=%s", a.pod.Name, a.pod.Status.PodIP)
```

Comment:
This panics when a.pod is nil:
https://travis-ci.org/linkerd/linkerd2/jobs/498829807#L1169

This is interesting. Other than this log line, `updateAddress.pod` is not required on the `remove` param. In the longer term we may want to consider modifying this method signature to be explicit about it:
```go
func (l *endpointListener) Update(add []*updateAddress, remove []*net.TcpAddress)
```

In the short term I suppose you could do a nil check here, or update the tests to always include pods in the `remove` slice, or both. My reading of the non-test code paths indicates we can always expect the pod to be present.

Either way it may be helpful to call this out in the top-line function comment.

Reply:
nice catch!